onnxruntime/core/providers/cpu/cpu_execution_provider.h

```diff
@@ -25,30 +25,17 @@ using FuseRuleFn = std::function<void(const onnxruntime::GraphViewer&,
 // Logical device representation.
 class CPUExecutionProvider : public IExecutionProvider {
  public:
-  explicit CPUExecutionProvider(const CPUExecutionProviderInfo& info)
-      : IExecutionProvider{onnxruntime::kCpuExecutionProvider} {
-    bool create_arena = info.create_arena;
-
-#if defined(USE_JEMALLOC) || defined(USE_MIMALLOC)
-    // JEMalloc/mimalloc already have memory pool, so just use device allocator.
-    create_arena = false;
-#elif !(defined(__amd64__) || defined(_M_AMD64) || defined(__aarch64__) || defined(_M_ARM64))
-    // Disable Arena allocator for x86_32 build because it may run into infinite loop when integer overflow happens
-    create_arena = false;
-#endif
-
-    AllocatorCreationInfo device_info{[](int) { return std::make_unique<CPUAllocator>(); },
-                                      DEFAULT_CPU_ALLOCATOR_DEVICE_ID, create_arena};
-
-    InsertAllocator(CreateAllocator(device_info));
-  }
+  // delay_allocator_registration = true is used to allow sharing of allocators between different providers that are
+  // associated with the same device
+  explicit CPUExecutionProvider(const CPUExecutionProviderInfo& info, bool delay_allocator_registration = false);
 
```

Comment:
Should we delay by default as InferenceSession is always going to call RegisterAllocator? 

I would have expected it's only in a unit test where you might create and use the CPU EP outside of an InferenceSession where you wouldn't want to do that. 

Or if you want to keep the default as false, does line 1234 in inference_session.cc which directly creates the default CPU EP without going via the factory need updating to delay?

Reply:
Updated line 1234. Yeah, I thought about the default being the other way but realized that there's just way too much dependency on the current behavior of CPU ep and these dependents don't use a uniform way to get an instance of the EP. So, for now I prefer keeping this default. Obviously, since this is internal to ORT we can change this decision later if and when we refactor the dependents.